server/mediaserver.go

```diff
@@ -515,50 +531,66 @@ func (s *LivepeerServer) registerConnection(ctx context.Context, rtmpStrm stream
 		Format:     params.Format,
 	}
 	hlsStrmID := core.MakeStreamID(mid, &vProfile)
-	s.connectionLock.RLock()
-	// Fast path - check early if session exists - creating new session can take time
-	oldCxn, exists := s.rtmpConnections[mid]
-	s.connectionLock.RUnlock()
+	playlist := core.NewBasicPlaylistManager(mid, storage, recordStorage)
+
+	// first, initialize connection without SessionManager, which creates O and T sessions, and may leave
+	// connectionLock locked for significant amount of time
+	cxn := &rtmpConnection{
+		mid:          mid,
+		initializing: make(chan error),
+		nonce:        params.Nonce,
+		stream:       rtmpStrm,
+		pl:           playlist,
+		profile:      &vProfile,
+		params:       params,
+		lastUsed:     time.Now(),
+	}
+	s.connectionLock.Lock()
+	oldCxn, exists := s.getActiveRtmpConnectionUnsafe(mid)
 	if exists {
 		// We can only have one concurrent stream per ManifestID
+		s.connectionLock.Unlock()
 		return oldCxn, errAlreadyExists
 	}
+	s.rtmpConnections[mid] = cxn
+	// do not obtain this lock again while initializing channel is open, it will cause deadlock if other goroutine already obtained the lock and called getActiveRtmpConnectionUnsafe()
+	s.connectionLock.Unlock()
 
-	playlist := core.NewBasicPlaylistManager(mid, storage, recordStorage)
+	// delete uninitialized connection in case of error
+	defer func() {
+		if <-cxn.initializing != nil {
```

Comment:
There doesn't seem to be a scenario where an error would actually be passed to `initializing` since it is only closed later on. If we end up just using `initializing` for signaling without any error cases that need to be communicated, I think the following would suffice:

```
<-cxn.initializing
s.connectionLock.Lock()
delete(s.rtmpConnections, mid)
s.connectionLock.Unlock()
```